File: platform/Flutter/sweeteditor/lib/completion/completion_popup_controller.dart
import 'completion_provider_manager.dart';
import 'completion_types.dart';
import '../overlay/editor_overlay.dart';

const int _maxVisibleItems = 6;
const double _itemHeightVp = 32;
const double _popupWidthVp = 300;
const double _gapVp = 4;

/// Popup position for the completion panel.
class PopupPosition {
  const PopupPosition({
    required this.x,
    required this.belowY,
    required this.aboveY,
    required this.popupWidth,
    required this.popupHeight,
  });

  final double x;
  final double belowY;
  final double aboveY;
  final double popupWidth;
  final double popupHeight;
}

/// Theme colors for the completion panel.
class CompletionThemeColors {
  const CompletionThemeColors({
    required this.panelBgColor,
    required this.panelBorderColor,
    required this.selectedBgColor,
    required this.labelColor,
    required this.detailColor,
  });

  final int panelBgColor;
  final int panelBorderColor;
  final int selectedBgColor;
  final int labelColor;
  final int detailColor;
}

class CompletionPopupOverlayState {
  const CompletionPopupOverlayState({
    required this.items,
    required this.selectedIndex,
    required this.position,
  });

  final List<CompletionItem> items;
  final int selectedIndex;
  final PopupPosition position;
}

/// Completion popup controller (logic layer).
class CompletionPopupController implements CompletionUpdateListener {
  CompletionPopupController({
    required int panelBgColor,
    required int panelBorderColor,
    required int selectedBgColor,
    required int labelColor,
    required int detailColor,
  }) : _panelBgColor = panelBgColor,
       _panelBorderColor = panelBorderColor,
       _selectedBgColor = selectedBgColor,
       _labelColor = labelColor,
       _detailColor = detailColor;

  List<CompletionItem> _items = [];
  int _selectedIndex = 0;
  bool _showing = false;
  void Function(CompletionItem)? _confirmHandler;
  double _cachedCursorX = 0;
  double _cachedCursorY = 0;
  double _cachedCursorHeight = 0;
  int _panelBgColor;
  int _panelBorderColor;
  int _selectedBgColor;
  int _labelColor;
  int _detailColor;
  EditorOverlayBinding<CompletionPopupOverlayState>? _overlayBinding;

  void applyTheme(
    int panelBgColor,
    int panelBorderColor,
    int selectedBgColor,
    int labelColor,
    int detailColor,
  ) {
    _panelBgColor = panelBgColor;
    _panelBorderColor = panelBorderColor;
    _selectedBgColor = selectedBgColor;
    _labelColor = labelColor;
    _detailColor = detailColor;
  }

  void setConfirmHandler(void Function(CompletionItem)? handler) {
    _confirmHandler = handler;
  }

  void setViewBuilder(CompletionItemViewBuilder? builder) {
    // Reserved for future custom rendering support.
  }

  void bindOverlay(EditorOverlayBinding<CompletionPopupOverlayState>? binding) {
    _overlayBinding = binding;
  }

  bool get isShowing => _showing;
  List<CompletionItem> get items => _items;
  int get selectedIndex => _selectedIndex;
  CompletionThemeColors get themeColors => CompletionThemeColors(
    panelBgColor: _panelBgColor,
    panelBorderColor: _panelBorderColor,
    selectedBgColor: _selectedBgColor,
    labelColor: _labelColor,
    detailColor: _detailColor,
  );

  @override
  void onCompletionItemsUpdated(List<CompletionItem> newItems) {
    _items = List.of(newItems);
    _selectedIndex = 0;
    if (_items.isEmpty) {
      dismiss();
    } else {
      _show();
    }
  }

  @override
  void onCompletionDismissed() => dismiss();

  bool handleKeyCode(int keyCode) {
    if (!_showing || _items.isEmpty) return false;
    switch (keyCode) {
      case 13: // Enter
        confirmSelected();
        return true;
      case 27: // Escape
        dismiss();
        return true;
      case 38: // Up
        _moveSelection(-1);
        return true;
      case 40: // Down
        _moveSelection(1);
        return true;
      default:
        return false;
    }
  }

  void updateCursorPosition(
    double cursorX,
    double cursorY,
    double cursorHeight,
  ) {
    _cachedCursorX = cursorX;
    _cachedCursorY = cursorY;
    _cachedCursorHeight = cursorHeight;
    if (_showing) {
      _overlayBinding?.update(_buildOverlayState());
    }
  }

  void dismiss() {
    if (!_showing) return;
    _showing = false;
    _overlayBinding?.hide();
  }

  void confirmSelected() {
    if (_selectedIndex >= 0 && _selectedIndex < _items.length) {
      final item = _items[_selectedIndex];
      dismiss();
      _confirmHandler?.call(item);
    }
  }

  void confirmItem(int index) {
    if (index >= 0 && index < _items.length) {
      _selectedIndex = index;
      confirmSelected();
    }
  }

  void _show() {
    final wasShowing = _showing;
    _showing = true;
    final overlayState = _buildOverlayState();
    if (wasShowing) {
      _overlayBinding?.update(overlayState);
    } else {
      _overlayBinding?.show(overlayState);
    }
  }

  void _moveSelection(int delta) {
    if (_items.isEmpty) return;
    final old = _selectedIndex;
    _selectedIndex = (_selectedIndex + delta).clamp(0, _items.length - 1);
    if (old != _selectedIndex) {
      _overlayBinding?.update(_buildOverlayState());
    }
  }

  CompletionPopupOverlayState _buildOverlayState() {
    return CompletionPopupOverlayState(
      items: List<CompletionItem>.unmodifiable(_items),
      selectedIndex: _selectedIndex,
      position: _computePosition(),
    );
  }

  PopupPosition _computePosition() {
    final popupHeight =
        _itemHeightVp * _items.length.clamp(0, _maxVisibleItems);
    final belowY = _cachedCursorY + _cachedCursorHeight + _gapVp;
    final aboveY = _cachedCursorY - popupHeight - _gapVp;
    return PopupPosition(
      x: _cachedCursorX,
      belowY: belowY,
      aboveY: aboveY,
      popupWidth: _popupWidthVp,
      popupHeight: popupHeight,
    );
  }
}

File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/completion/CompletionPopupController.java
package com.qiplat.sweeteditor.completion;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.GradientDrawable;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.PopupWindow;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.qiplat.sweeteditor.EditorTheme;
import com.qiplat.sweeteditor.animation.PopupAnimator;
import com.qiplat.sweeteditor.popup.PopupPositioner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Completion popup controller: PopupWindow + RecyclerView.
 * <p>Cursor-following positioning, up/down key navigation, Enter to confirm, Escape to dismiss.</p>
 */
public class CompletionPopupController implements CompletionProviderManager.CompletionUpdateListener {

    public interface CompletionConfirmListener {
        void onCompletionConfirmed(@NonNull CompletionItem item);
    }

    private static final int MAX_VISIBLE_ITEMS = 4;
    private static final int ITEM_HEIGHT_DP = 32;
    private static final int POPUP_WIDTH_DP = 300;
    private static final int PANEL_HORIZONTAL_PADDING_DP = 4;
    private static final int PANEL_VERTICAL_PADDING_DP = 6;
    private static final int POPUP_MIN_WIDTH_DP = 120;
    private static final int POPUP_FRAME_PADDING_DP = 8;
    private static final int GAP_DP = 4;
    private static final List<PopupPositioner.Placement> POPUP_PLACEMENTS = Arrays.asList(
            PopupPositioner.Placement.of(PopupPositioner.PopupSide.BELOW, PopupPositioner.PopupAlign.START),
            PopupPositioner.Placement.of(PopupPositioner.PopupSide.ABOVE, PopupPositioner.PopupAlign.START)
    );

    private final Context context;
    private final View anchorView;
    @Nullable private CompletionConfirmListener confirmListener;
    @Nullable private CompletionItemViewFactory viewFactory;

    private PopupWindow popupWindow;
    private RecyclerView recyclerView;
    private CompletionAdapter adapter;
    private final List<CompletionItem> items = new ArrayList<>();
    private int selectedIndex = 0;

    private int panelBgColor;
    private int panelBorderColor;
    private int selectedBgColor;
    private int labelColor;
    private int detailColor;

    private float cachedCursorX = 0;
    private float cachedCursorY = 0;
    private float cachedCursorHeight = 0;
    @NonNull private PopupPositioner.Placement lastPlacement =
            PopupPositioner.Placement.of(PopupPositioner.PopupSide.BELOW, PopupPositioner.PopupAlign.START);

    public CompletionPopupController(@NonNull Context context, @NonNull View anchorView, @NonNull EditorTheme theme) {
        this.context = context;
        this.anchorView = anchorView;
        panelBgColor = theme.completionBgColor;
        panelBorderColor = theme.completionBorderColor;
        selectedBgColor = theme.completionSelectedBgColor;
        labelColor = theme.completionLabelColor;
        detailColor = theme.completionDetailColor;
        initPopup();
    }

    public void applyTheme(@NonNull EditorTheme theme) {
        panelBgColor = theme.completionBgColor;
        panelBorderColor = theme.completionBorderColor;
        selectedBgColor = theme.completionSelectedBgColor;
        labelColor = theme.completionLabelColor;
        detailColor = theme.completionDetailColor;
        if (recyclerView != null) {
            recyclerView.setBackground(createPanelBackground());
            recyclerView.setClipToOutline(true);
        }
        if (adapter != null) adapter.notifyDataSetChanged();
    }

    public void setConfirmListener(@Nullable CompletionConfirmListener listener) {
        this.confirmListener = listener;
    }

    public void setViewFactory(@Nullable CompletionItemViewFactory factory) {
        this.viewFactory = factory;
        if (adapter != null) adapter.setViewFactory(factory);
    }

    public boolean isShowing() {
        return popupWindow != null && popupWindow.isShowing();
    }

    @Override
    public void onCompletionItemsUpdated(@NonNull List<CompletionItem> newItems) {
        items.clear();
        items.addAll(newItems);
        selectedIndex = 0;
        adapter.notifyDataSetChanged();
        if (items.isEmpty()) {
            dismiss();
        } else {
            show();
        }
    }

    @Override
    public void onCompletionDismissed() {
        dismiss();
    }

    /**
     * Handle Android KeyEvent keyCode for completion panel navigation.
     */
    public boolean handleAndroidKeyCode(int androidKeyCode) {
        if (!isShowing() || items.isEmpty()) return false;
        switch (androidKeyCode) {
            case KeyEvent.KEYCODE_ENTER:
                confirmSelected();
                return true;
            case KeyEvent.KEYCODE_ESCAPE:
                dismiss();
                return true;
            case KeyEvent.KEYCODE_DPAD_UP:
                moveSelection(-1);
                return true;
            case KeyEvent.KEYCODE_DPAD_DOWN:
                moveSelection(1);
                return true;
            default:
                return false;
        }
    }

    /**
     * Update cached cursor screen coordinates (called by SweetEditor every frame in onDraw).
     * If panel is showing, also refresh panel position.
     */
    public void updateCursorPosition(float cursorScreenX, float cursorScreenY, float cursorHeight) {
        cachedCursorX = cursorScreenX;
        cachedCursorY = cursorScreenY;
        cachedCursorHeight = cursorHeight;
        if (isShowing()) {
            applyPosition();
        }
    }

    /**
     * Calculate and apply panel position based on cached cursor coordinates.
     * cachedCursorX/Y are coordinates within anchorView, need to be converted to screen coordinates for PopupWindow positioning.
     */
    private void applyPosition() {
        PopupLayout layout = computePopupLayout();
        if (layout == null) {
            return;
        }
        applyPopupLayout(layout);
    }

    public void dismiss() {
        if (popupWindow != null && popupWindow.isShowing()) {
            PopupAnimator.animateDismiss(recyclerView, lastPlacement, () -> {
                if (popupWindow.isShowing()) {
                    popupWindow.dismiss();
                }
            });
        }
    }

    private void initPopup() {
        recyclerView = new RecyclerView(context);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));

        recyclerView.setBackground(createPanelBackground());
        recyclerView.setClipToOutline(true);
        recyclerView.setPadding(
                PopupPositioner.dpToPx(context, PANEL_HORIZONTAL_PADDING_DP),
                PopupPositioner.dpToPx(context, PANEL_VERTICAL_PADDING_DP),
                PopupPositioner.dpToPx(context, PANEL_HORIZONTAL_PADDING_DP),
                PopupPositioner.dpToPx(context, PANEL_VERTICAL_PADDING_DP));
        recyclerView.setClipToPadding(false);

        adapter = new CompletionAdapter();
        recyclerView.setAdapter(adapter);

        int width = PopupPositioner.dpToPx(context, POPUP_WIDTH_DP);
        popupWindow = new PopupWindow(recyclerView, width, ViewGroup.LayoutParams.WRAP_CONTENT);
        popupWindow.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        popupWindow.setFocusable(false);
        popupWindow.setElevation(PopupPositioner.dpToPx(context, 8));
    }

    private void show() {
        PopupLayout layout = computePopupLayout();
        if (layout == null) {
            dismiss();
            return;
        }
        applyPopupSize(layout.popupWidth, layout.popupHeight);
        if (!popupWindow.isShowing()) {
            lastPlacement = layout.position.placement;
            PopupAnimator.prepareForShow(recyclerView, lastPlacement);
            popupWindow.showAtLocation(anchorView, Gravity.NO_GRAVITY,
                    layout.position.screenX, layout.position.screenY);
            applyPopupLayout(layout);
            PopupAnimator.animateShow(recyclerView, lastPlacement);
            return;
        }
        applyPopupLayout(layout);
    }

    @Nullable
    private PopupLayout computePopupLayout() {
        int gap = PopupPositioner.dpToPx(context, GAP_DP);
        int visibleRows = Math.max(1, Math.min(items.size(), MAX_VISIBLE_ITEMS));
        int popupHeight = PopupPositioner.dpToPx(context,
                ITEM_HEIGHT_DP * visibleRows + PANEL_VERTICAL_PADDING_DP * 2);

        Rect availableFrame = PopupPositioner.resolveAvailableFrame(anchorView);
        if (availableFrame.height() < PopupPositioner.dpToPx(context, ITEM_HEIGHT_DP)) {
            return null;
        }

        int desiredPopupWidth = PopupPositioner.dpToPx(context, POPUP_WIDTH_DP);
        int minPopupWidth = PopupPositioner.dpToPx(context, POPUP_MIN_WIDTH_DP);
        int availableWidth = Math.max(1, availableFrame.width() - PopupPositioner.dpToPx(context, POPUP_FRAME_PADDING_DP));
        int popupWidth = Math.min(desiredPopupWidth, Math.max(minPopupWidth, availableWidth));
        if (popupWidth > availableFrame.width()) {
            popupWidth = availableFrame.width();
        }

        int maxPopupHeight = Math.max(
                PopupPositioner.dpToPx(context, ITEM_HEIGHT_DP + PANEL_VERTICAL_PADDING_DP * 2),
                availableFrame.height() - PopupPositioner.dpToPx(context, POPUP_FRAME_PADDING_DP));
        if (popupHeight > maxPopupHeight) {
            popupHeight = maxPopupHeight;
        }

        PopupPositioner.Result position = PopupPositioner.compute(new PopupPositioner.Request(
                anchorView,
                new RectF(cachedCursorX, cachedCursorY, cachedCursorX, cachedCursorY + cachedCursorHeight),
                popupWidth,
                popupHeight,
                gap,
                0,
                POPUP_PLACEMENTS
        ));
        return new PopupLayout(position, popupWidth, popupHeight);
    }

    private void applyPopupSize(int popupWidth, int popupHeight) {
        ViewGroup.LayoutParams layoutParams = recyclerView.getLayoutParams();
        if (layoutParams == null) {
            layoutParams = new ViewGroup.LayoutParams(popupWidth, popupHeight);
        } else {
            layoutParams.width = popupWidth;
            layoutParams.height = popupHeight;
        }
        recyclerView.setLayoutParams(layoutParams);
        popupWindow.setWidth(popupWidth);
        popupWindow.setHeight(popupHeight);
    }

    private void applyPopupLayout(@NonNull PopupLayout layout) {
        lastPlacement = layout.position.placement;
        applyPopupSize(layout.popupWidth, layout.popupHeight);
        popupWindow.update(layout.position.screenX, layout.position.screenY,
                layout.popupWidth, layout.popupHeight);
    }

    @NonNull
    private GradientDrawable createPanelBackground() {
        GradientDrawable panelBg = new GradientDrawable();
        panelBg.setColor(panelBgColor);
        panelBg.setCornerRadius(PopupPositioner.dpToPx(context, 12));
        panelBg.setStroke(PopupPositioner.dpToPx(context, 1), panelBorderColor);
        return panelBg;
    }

    private void moveSelection(int delta) {
        if (items.isEmpty()) return;
        int old = selectedIndex;
        selectedIndex = Math.max(0, Math.min(items.size() - 1, selectedIndex + delta));
        if (old != selectedIndex) {
            adapter.notifyItemChanged(old);
            adapter.notifyItemChanged(selectedIndex);
            recyclerView.scrollToPosition(selectedIndex);
        }
    }

    private void confirmSelected() {
        if (selectedIndex >= 0 && selectedIndex < items.size()) {
            CompletionItem item = items.get(selectedIndex);
            dismiss();
            if (confirmListener != null) {
                confirmListener.onCompletionConfirmed(item);
            }
        }
    }

    private class CompletionAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

        @Nullable private CompletionItemViewFactory factory;

        void setViewFactory(@Nullable CompletionItemViewFactory factory) {
            this.factory = factory;
        }

        @NonNull @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            if (factory != null) {
                View view = factory.createItemView(parent);
                return new RecyclerView.ViewHolder(view) {};
            }
            return new DefaultViewHolder(parent);
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
            CompletionItem item = items.get(holder.getAdapterPosition());
            boolean isSelected = position == selectedIndex;
            if (factory != null) {
                factory.bindItemView(holder.itemView, item, isSelected);
            } else {
                ((DefaultViewHolder) holder).bind(item, isSelected, selectedBgColor, labelColor, detailColor);
            }
            holder.itemView.setOnClickListener(v -> {
                selectedIndex = position;
                confirmSelected();
            });
        }

        @Override
        public int getItemCount() {
            return items.size();
        }
    }

    private static class DefaultViewHolder extends RecyclerView.ViewHolder {
        private final TextView kindBadge;
        private final ImageView iconView;
        private final TextView labelView;
        private final TextView detailView;
        private final GradientDrawable rowBg;
        private final GradientDrawable badgeBg;

        DefaultViewHolder(@NonNull ViewGroup parent) {
            super(createDefaultItemView(parent.getContext()));
            kindBadge = itemView.findViewWithTag("kindBadge");
            iconView = itemView.findViewWithTag("icon");
            labelView = itemView.findViewWithTag("label");
            detailView = itemView.findViewWithTag("detail");

            rowBg = new GradientDrawable();
            rowBg.setCornerRadius(PopupPositioner.dpToPx(parent.getContext(), 6));
            itemView.setBackground(rowBg);

            badgeBg = new GradientDrawable();
            badgeBg.setCornerRadius(PopupPositioner.dpToPx(parent.getContext(), 4));
            kindBadge.setBackground(badgeBg);
        }

        void bind(@NonNull CompletionItem item, boolean isSelected, int selectedColor, int lblColor, int dtlColor) {
            labelView.setText(item.label);
            labelView.setTextColor(lblColor);
            if (item.detail != null && !item.detail.isEmpty()) {
                detailView.setVisibility(View.VISIBLE);
                detailView.setText(item.detail);
                detailView.setTextColor(dtlColor);
            } else {
                detailView.setVisibility(View.GONE);
            }

            rowBg.setColor(isSelected ? selectedColor : Color.TRANSPARENT);

            iconView.setVisibility(View.GONE);
            kindBadge.setVisibility(View.VISIBLE);
            applyKindBadge(kindBadge, item.kind);
        }

        private static void applyKindBadge(TextView badge, int kind) {
            int color;
            String letter;
            switch (kind) {
                case CompletionItem.KIND_KEYWORD:
                    color = 0xFFC678DD; letter = "K"; break;
                case CompletionItem.KIND_FUNCTION:
                    color = 0xFF61AFEF; letter = "F"; break;
                case CompletionItem.KIND_VARIABLE:
                    color = 0xFFE5C07B; letter = "V"; break;
                case CompletionItem.KIND_CLASS:
                    color = 0xFFE06C75; letter = "C"; break;
                case CompletionItem.KIND_INTERFACE:
                    color = 0xFF56B6C2; letter = "I"; break;
                case CompletionItem.KIND_MODULE:
                    color = 0xFFD19A66; letter = "M"; break;
                case CompletionItem.KIND_PROPERTY:
                    color = 0xFF98C379; letter = "P"; break;
                case CompletionItem.KIND_SNIPPET:
                    color = 0xFFBE5046; letter = "S"; break;
                default:
                    color = 0xFF7A8494; letter = "T"; break;
            }
            badge.setText(letter);
            GradientDrawable bg = (GradientDrawable) badge.getBackground();
            bg.setColor(color);
        }

        private static View createDefaultItemView(@NonNull Context context) {
            float density = context.getResources().getDisplayMetrics().density;
            int hPadding = (int) (8 * density);
            int vPadding = (int) (2 * density);
            int height = (int) (ITEM_HEIGHT_DP * density);

            android.widget.LinearLayout layout = new android.widget.LinearLayout(context);
            layout.setOrientation(android.widget.LinearLayout.HORIZONTAL);
            layout.setGravity(Gravity.CENTER_VERTICAL);
            layout.setPadding(hPadding, vPadding, hPadding, vPadding);
            layout.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, height));

            TextView kindBadge = new TextView(context);
            int badgeSize = (int) (18 * density);
            android.widget.LinearLayout.LayoutParams badgeLp =
                    new android.widget.LinearLayout.LayoutParams(badgeSize, badgeSize);
            badgeLp.setMarginEnd((int) (8 * density));
            kindBadge.setLayoutParams(badgeLp);
            kindBadge.setGravity(Gravity.CENTER);
            kindBadge.setTextSize(10);
            kindBadge.setTextColor(0xFFFFFFFF);
            kindBadge.setTypeface(Typeface.DEFAULT_BOLD);
            kindBadge.setTag("kindBadge");
            layout.addView(kindBadge);

            ImageView icon = new ImageView(context);
            int iconSize = (int) (16 * density);
            android.widget.LinearLayout.LayoutParams iconLp =
                    new android.widget.LinearLayout.LayoutParams(iconSize, iconSize);
            iconLp.setMarginEnd((int) (8 * density));
            icon.setLayoutParams(iconLp);
            icon.setTag("icon");
            icon.setVisibility(View.GONE);
            layout.addView(icon);

            TextView label = new TextView(context);
            label.setTextSize(13);
            label.setTextColor(0xFFD8DEE9);
            label.setSingleLine(true);
            label.setEllipsize(TextUtils.TruncateAt.END);
            label.setTag("label");
            android.widget.LinearLayout.LayoutParams labelLp =
                    new android.widget.LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1);
            label.setLayoutParams(labelLp);
            layout.addView(label);

            TextView detail = new TextView(context);
            detail.setTextSize(11);
            detail.setTextColor(0xFF7A8494);
            detail.setSingleLine(true);
            detail.setEllipsize(TextUtils.TruncateAt.END);
            detail.setTag("detail");
            detail.setVisibility(View.GONE);
            android.widget.LinearLayout.LayoutParams detailLp =
                    new android.widget.LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            detailLp.setMarginStart((int) (8 * density));
            detail.setLayoutParams(detailLp);
            layout.addView(detail);

            return layout;
        }
    }

    private static final class PopupLayout {
        @NonNull final PopupPositioner.Result position;
        final int popupWidth;
        final int popupHeight;

        PopupLayout(@NonNull PopupPositioner.Result position, int popupWidth, int popupHeight) {
            this.position = position;
            this.popupWidth = popupWidth;
            this.popupHeight = popupHeight;
        }
    }
}
